{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Small_Molecule Dataset Download and Preprocessing "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Preprocessing data code from repository suinleelab **contrastiveVI**: https://github.com/suinleelab/contrastiveVI/blob/main/contrastive_vi/data/datasets/mcfarland_2020.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Small Molecule dataset is collected by McFarland et al. (2020) in paper \"Multiplexed single-cell transcriptional response profiling to define cancer vulnerabilities and therapeutic mechanism of action\"(https://www.nature.com/articles/s41467-020-17440-w).\n",
    "\n",
    "There is no need to find the data to download since the downoad link is embedded in the download function. Nonetheless, the code is available at https://figshare.com/s/139f64b495dea9d88c70. We specifically look at data zip file (\"Idasanutlin_24hr_expt1\", and \"DMSO_24hr_expt1\"). The data is licensed under CC BY 4.0 License. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The python **anndata** package is found here: https://github.com/scverse/anndata?tab=BSD-3-Clause-1-ov-file\n",
    "\n",
    "The **scanpy** package is found here: https://github.com/scverse/scanpy\n",
    "\n",
    "The **requests** package is found here: https://github.com/psf/requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Download, read, and preprocess Mcfarland et al. (2020) expression data.\n",
    "\n",
    "Single-cell expression data from Mcfarland et al. Multiplexed single-cell\n",
    "transcriptional response profiling to define cancer vulnerabilities and therapeutic\n",
    "mechanism of action. Nature Communications (2020).\n",
    "\"\"\"\n",
    "import os\n",
    "import shutil\n",
    "from typing import Tuple\n",
    "\n",
    "import anndata\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from anndata import AnnData\n",
    "from scipy.io import mmread\n",
    "import requests\n",
    "\n",
    "\n",
    "def download_binary_file(file_url: str, output_path: str) -> None:\n",
    "    \"\"\"\n",
    "    Download binary data file from a URL.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        file_url: URL where the file is hosted.\n",
    "        output_path: Output path for the downloaded file.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        None.\n",
    "    \"\"\"\n",
    "    request = requests.get(file_url)\n",
    "    with open(output_path, \"wb\") as f:\n",
    "        f.write(request.content)\n",
    "    print(f\"Downloaded data from {file_url} at {output_path}\")\n",
    "\n",
    "    \n",
    "    \n",
    "def download_mcfarland_2020(output_path: str) -> None:\n",
    "    \"\"\"\n",
    "    Download Mcfarland et al. 2020 data from the hosting URLs.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        output_path: Output path to store the downloaded and unzipped\n",
    "        directories.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        None. File directories are downloaded and unzipped in output_path.\n",
    "    \"\"\"\n",
    "    idasanutlin_url = \"https://figshare.com/ndownloader/files/18716351\"\n",
    "    idasanutlin_output_filename = os.path.join(output_path, \"idasanutlin.zip\")\n",
    "\n",
    "    download_binary_file(idasanutlin_url, idasanutlin_output_filename)\n",
    "    idasanutlin_output_dir = idasanutlin_output_filename.replace(\".zip\", \"\")\n",
    "    shutil.unpack_archive(idasanutlin_output_filename, idasanutlin_output_dir)\n",
    "\n",
    "    dmso_url = \"https://figshare.com/ndownloader/files/18716354\"\n",
    "    dmso_output_filename = os.path.join(output_path, \"dmso.zip\")\n",
    "\n",
    "    download_binary_file(dmso_url, dmso_output_filename)\n",
    "    dmso_output_dir = dmso_output_filename.replace(\".zip\", \"\")\n",
    "    shutil.unpack_archive(dmso_output_filename, dmso_output_dir)\n",
    "\n",
    "\n",
    "def _read_mixseq_df(directory: str) -> pd.DataFrame:\n",
    "    data = mmread(os.path.join(directory, \"matrix.mtx\"))\n",
    "    barcodes = pd.read_table(os.path.join(directory, \"barcodes.tsv\"), header=None)\n",
    "    classifications = pd.read_csv(os.path.join(directory, \"classifications.csv\"))\n",
    "    classifications[\"cell_line\"] = np.array(\n",
    "        [x.split(\"_\")[0] for x in classifications.singlet_ID.values]\n",
    "    )\n",
    "    gene_names = pd.read_table(os.path.join(directory, \"genes.tsv\"), header=None)\n",
    "\n",
    "    df = pd.DataFrame(\n",
    "        data.toarray(),\n",
    "        columns=barcodes.iloc[:, 0].values,\n",
    "        index=gene_names.iloc[:, 0].values,\n",
    "    )\n",
    "    return df\n",
    "\n",
    "\n",
    "def _get_tp53_mutation_status(directory: str) -> np.array:\n",
    "    # Taken from https://cancerdatascience.org/blog/posts/mix-seq/\n",
    "    TP53_WT = [\n",
    "        \"LNCAPCLONEFGC_PROSTATE\",\n",
    "        \"DKMG_CENTRAL_NERVOUS_SYSTEM\",\n",
    "        \"NCIH226_LUNG\",\n",
    "        \"RCC10RGB_KIDNEY\",\n",
    "        \"SNU1079_BILIARY_TRACT\",\n",
    "        \"CCFSTTG1_CENTRAL_NERVOUS_SYSTEM\",\n",
    "        \"COV434_OVARY\",\n",
    "    ]\n",
    "\n",
    "    classifications = pd.read_csv(os.path.join(directory, \"classifications.csv\"))\n",
    "    TP53_mutation_status = [\n",
    "        \"Wild Type\" if x in TP53_WT else \"Mutation\"\n",
    "        for x in classifications.singlet_ID.values\n",
    "    ]\n",
    "    return np.array(TP53_mutation_status)\n",
    "\n",
    "\n",
    "def read_mcfarland_2020(file_directory: str) -> Tuple[pd.DataFrame, pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Read the expression data for Mcfarland et al. 2020 in the given directory.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        file_directory: Directory containing Mcfarland et al. 2020 data.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        Two data frames of raw count expression data. The first contains\n",
    "        single-cell gene expression count data from cancer cell lines exposed to\n",
    "        idasanutlin with cell identification barcodes as column names and gene IDs as\n",
    "        indices. The second contains count data with the same format from samples\n",
    "        exposed to a control solution (DMSO).\n",
    "    \"\"\"\n",
    "    idasanutlin_dir = os.path.join(\n",
    "        file_directory, \"idasanutlin\", \"Idasanutlin_24hr_expt1\"\n",
    "    )\n",
    "    idasanutlin_df = _read_mixseq_df(idasanutlin_dir)\n",
    "\n",
    "    dmso_dir = os.path.join(file_directory, \"dmso\", \"DMSO_24hr_expt1\")\n",
    "    dmso_df = _read_mixseq_df(dmso_dir)\n",
    "\n",
    "    return idasanutlin_df, dmso_df\n",
    "\n",
    "\n",
    "def preprocess_mcfarland_2020(download_path: str, n_top_genes: int) -> AnnData:\n",
    "    \"\"\"\n",
    "    Preprocess expression data from Mcfarland et al., 2020.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        download_path: Path containing the downloaded Mcfarland et al. 2020 data files.\n",
    "        n_top_genes: Number of most variable genes to retain.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        An AnnData object containing single-cell expression data. The layer\n",
    "        \"count\" contains the count data for the most variable genes. The X\n",
    "        variable contains the total-count-normalized and log-transformed data\n",
    "        for the most variable genes (a copy with all the genes is stored in\n",
    "        .raw).\n",
    "    \"\"\"\n",
    "\n",
    "    idasanutlin_df, dmso_df = read_mcfarland_2020(download_path)\n",
    "    idasanutlin_df, dmso_df = idasanutlin_df.transpose(), dmso_df.transpose()\n",
    "\n",
    "    m, n = idasanutlin_df.shape\n",
    "    p, k = dmso_df.shape\n",
    "\n",
    "\n",
    "    idasanutlin_adata = AnnData(idasanutlin_df)\n",
    "    idasanutlin_dir = os.path.join(\n",
    "        download_path, \"idasanutlin\", \"Idasanutlin_24hr_expt1\"\n",
    "    )\n",
    "    idasanutlin_adata.obs[\"TP53_mutation_status\"] = _get_tp53_mutation_status(\n",
    "        idasanutlin_dir\n",
    "    )\n",
    "    idasanutlin_adata.obs[\"condition\"] = np.repeat(\n",
    "        \"Idasanutlin\", idasanutlin_adata.shape[0]\n",
    "    )\n",
    "\n",
    "    dmso_adata = AnnData(dmso_df)\n",
    "    dmso_dir = os.path.join(download_path, \"dmso\", \"DMSO_24hr_expt1\")\n",
    "    dmso_adata.obs[\"TP53_mutation_status\"] = _get_tp53_mutation_status(dmso_dir)\n",
    "    dmso_adata.obs[\"condition\"] = np.repeat(\"DMSO\", dmso_adata.shape[0])\n",
    "\n",
    "    full_adata = anndata.concat([idasanutlin_adata, dmso_adata])\n",
    "    full_adata.layers[\"count\"] = full_adata.X.copy()\n",
    "    sc.pp.normalize_total(full_adata)\n",
    "    sc.pp.log1p(full_adata)\n",
    "    full_adata.raw = full_adata\n",
    "    sc.pp.highly_variable_genes(\n",
    "        full_adata,\n",
    "        flavor=\"seurat_v3\",\n",
    "        n_top_genes=n_top_genes,\n",
    "        layer=\"count\",\n",
    "        subset=True,\n",
    "    )\n",
    "    full_adata = full_adata[\n",
    "        full_adata.layers[\"count\"].sum(1) != 0\n",
    "    ]  # Remove cells with all zeros.\n",
    "\n",
    "    full_data = full_adata.to_df()\n",
    "\n",
    "    return full_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "root_data_path = \"...\"   # local computer data path to downloaad the dataset. \n",
    "\n",
    "data = preprocess_mcfarland_2020(root_data_path, 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "idasanutlin_df row: 3097\n",
    "dmso_df row: 2831"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5927\n"
     ]
    }
   ],
   "source": [
    "n, p = data.shape\n",
    "print(n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "foreground = data[:3096, ]\n",
    "background = data[3097:, ]\n",
    "\n",
    "foreground = foreground.to_df()\n",
    "background = background.to_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "foreground.to_csv(\"small_molecule_fore_1.csv\")\n",
    "background.to_csv(\"small_molecule_back_1.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
