{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BMMC Data Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Preprocessing data code from repository suinleelab **contrastiveVI**: \"https://github.com/suinleelab/contrastiveVI/blob/main/contrastive_vi/data/datasets/zheng_2017.py\", and andrewcharlesjones **papca**: https://github.com/andrewcharlesjones/pcpca/blob/main/data/preprocess_singlecell_data.py\n",
    "\n",
    "And the function (get_data_df) is from pcpca experiment preprocessing file: \n",
    " https://github.com/andrewcharlesjones/pcpca/blob/main/data/preprocess_singlecell_data.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The python **anndata** package is found here: https://github.com/scverse/anndata?tab=BSD-3-Clause-1-ov-file\n",
    "\n",
    "The **scanpy** package is found here: https://github.com/scverse/scanpy\n",
    "\n",
    "The **requests** package is found here: https://github.com/psf/requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Download, read, and preprocess Zheng et al. (2017) expression data.\n",
    "\n",
    "Single-cell expression data from Zheng et al. Massively parallel digital\n",
    "transcriptional profiling of single cells. Nature Communications (2017).\n",
    "\"\"\"\n",
    "import os\n",
    "import shutil\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from anndata import AnnData\n",
    "from scipy.io import mmread\n",
    "import requests\n",
    "import seaborn as sns\n",
    "from os.path import join as pjoin\n",
    "from scipy.io import mmread\n",
    "\n",
    "N_GENES = 500\n",
    "\n",
    "\n",
    "# create a folder name \"singlecell_bmmc\" in local computer\n",
    "DATA_DIR = \".../singlecell_bmmc\"\n",
    "\n",
    "\n",
    "# this following function (download_binary_file) is from contrastive_vi.data.utils\n",
    "# (https://github.com/suinleelab/contrastiveVI/blob/main/contrastive_vi/data/utils.py) \n",
    "def download_binary_file(file_url: str, output_path: str) -> None:\n",
    "    \"\"\"\n",
    "    Download binary data file from a URL.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        file_url: URL where the file is hosted.\n",
    "        output_path: Output path for the downloaded file.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        None.\n",
    "    \"\"\"\n",
    "    request = requests.get(file_url)\n",
    "    with open(output_path, \"wb\") as f:\n",
    "        f.write(request.content)\n",
    "    print(f\"Downloaded data from {file_url} at {output_path}\")\n",
    "\n",
    "\n",
    "def download_zheng_2017(output_path: str) -> None:\n",
    "    \"\"\"\n",
    "    Download Zheng et al. 2017 data from the hosting URLs.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        output_path: Output path to store the downloaded and unzipped\n",
    "        directories.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        None. File directories are downloaded and unzipped in output_path.\n",
    "    \"\"\"\n",
    "    host = \"https://cf.10xgenomics.com/samples/cell-exp/1.1.0/\"\n",
    "    host_directories = [\n",
    "        (\n",
    "            \"aml027_post_transplant/\"\n",
    "            \"aml027_post_transplant_filtered_gene_bc_matrices.tar.gz\"\n",
    "        ),\n",
    "        (\n",
    "            \"aml027_pre_transplant/\"\n",
    "            \"aml027_pre_transplant_filtered_gene_bc_matrices.tar.gz\"\n",
    "        ),\n",
    "        (\n",
    "            \"aml035_post_transplant/\"\n",
    "            \"aml035_post_transplant_filtered_gene_bc_matrices.tar.gz\"\n",
    "        ),\n",
    "        (\n",
    "            \"aml035_pre_transplant/\"\n",
    "            \"aml035_pre_transplant_filtered_gene_bc_matrices.tar.gz\"\n",
    "        ),\n",
    "        (\n",
    "            \"frozen_bmmc_healthy_donor1/\"\n",
    "            \"frozen_bmmc_healthy_donor1_filtered_gene_bc_matrices.tar.gz\"\n",
    "        ),\n",
    "        (\n",
    "            \"frozen_bmmc_healthy_donor2/\"\n",
    "            \"frozen_bmmc_healthy_donor2_filtered_gene_bc_matrices.tar.gz\"\n",
    "        ),\n",
    "    ]\n",
    "    urls = [host + host_directory for host_directory in host_directories]\n",
    "    output_filenames = [os.path.join(output_path, url.split(\"/\")[-1]) for url in urls]\n",
    "    for url, output_filename in zip(urls, output_filenames):\n",
    "        download_binary_file(url, output_filename)\n",
    "        output_dir = output_filename.replace(\".tar.gz\", \"\")\n",
    "        shutil.unpack_archive(output_filename, output_dir)\n",
    "\n",
    "\n",
    "# this following function (get_data_df) is from pcpca experiment preprocessing file: \n",
    "# https://github.com/andrewcharlesjones/pcpca/blob/main/data/preprocess_singlecell_data.py\n",
    "# the directorry to the data is modified becaause the two function above made packages in the root directory. \n",
    "def get_data_df(file_directory: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Read the expression data for in a downloaded file directory.\n",
    "\n",
    "    Args:\n",
    "    ----\n",
    "        file_directory: A downloaded and unzipped file directory.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        A data frame containing single-cell gene expression count, with cell\n",
    "        identification barcodes as column names and gene IDs as indices.\n",
    "    \"\"\"\n",
    "    data = mmread(\n",
    "        os.path.join(file_directory, \"filtered_matrices_mex/hg19/matrix.mtx\")\n",
    "    ).toarray()\n",
    "\n",
    "    # this line is from pcpca\n",
    "    data = np.log(data + 1)\n",
    "\n",
    "    genes = pd.read_table(os.path.join(file_directory, \"filtered_matrices_mex/hg19/genes.tsv\"), header=None)\n",
    "    barcodes = pd.read_table(os.path.join(file_directory, \"filtered_matrices_mex/hg19/barcodes.tsv\"), header=None)\n",
    "\n",
    "    # this following line is from pcpca\n",
    "    data_df = pd.DataFrame(\n",
    "        data, index=genes.iloc[:, 0].values, columns=barcodes.iloc[:, 0].values\n",
    "    )\n",
    "\n",
    "    data_df = data_df.iloc[:, np.sum(data_df.values, axis=0) != 0]\n",
    "    data_df = data_df.iloc[np.sum(data_df.values, axis=1) != 0, :]\n",
    "    return data_df.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "root_data_path = \"...\"   # local computer data path to download the dataset. \n",
    "\n",
    "download_zheng_2017(root_data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# preprocess the single cell data\n",
    "\n",
    "pretransplant1 = get_data_df(pjoin(DATA_DIR, \"aml027_pre_transplant_filtered_gene_bc_matrices\"))\n",
    "posttransplant1 = get_data_df(pjoin(DATA_DIR, \"aml027_post_transplant_filtered_gene_bc_matrices\"))\n",
    "pretransplant2 = get_data_df(pjoin(DATA_DIR, \"aml035_pre_transplant_filtered_gene_bc_matrices\"))\n",
    "posttransplant2 = get_data_df(pjoin(DATA_DIR, \"aml035_post_transplant_filtered_gene_bc_matrices\"))\n",
    "\n",
    "healthy1 = get_data_df(pjoin(DATA_DIR, \"frozen_bmmc_healthy_donor1_filtered_gene_bc_matrices\"))\n",
    "healthy2 = get_data_df(pjoin(DATA_DIR, \"frozen_bmmc_healthy_donor2_filtered_gene_bc_matrices\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total of 12399 cells and 12079 genes\n",
      "Saving 500 genes\n"
     ]
    }
   ],
   "source": [
    "# this following function (get_data_df) is from pcpca experiment preprocessing file: \n",
    "# https://github.com/andrewcharlesjones/pcpca/blob/main/data/preprocess_singlecell_data.py\n",
    "\n",
    "## Subset to shared genes\n",
    "shared_genes = pretransplant1.columns.values\n",
    "for curr_df in [posttransplant1, pretransplant2, posttransplant2, healthy1, healthy2]:\n",
    "    shared_genes = np.intersect1d(shared_genes, curr_df.columns.values)\n",
    "\n",
    "## Combine into one dataframe\n",
    "stacked_df = pretransplant1[shared_genes]\n",
    "for curr_df in [\n",
    "    posttransplant1,\n",
    "    pretransplant2,\n",
    "    posttransplant2,\n",
    "]:  # , healthy1, healthy2]:\n",
    "    stacked_df = pd.concat([stacked_df, curr_df[shared_genes]], axis=0)\n",
    "print(\"Total of {} cells and {} genes\".format(stacked_df.shape[0], stacked_df.shape[1]))\n",
    "\n",
    "## Subset to most variable genes\n",
    "gene_means = np.mean(stacked_df.values, axis=0)\n",
    "gene_vars = np.var(stacked_df.values, axis=0)\n",
    "gene_dispersions = gene_vars / gene_means\n",
    "top_idx = np.argsort(-gene_dispersions)[:N_GENES]\n",
    "top_genes = stacked_df.columns.values[top_idx]\n",
    "\n",
    "print(\"Saving {} genes\".format(top_genes.shape[0]))\n",
    "\n",
    "\n",
    "## Save\n",
    "pretransplant1[top_genes].to_csv(pjoin(DATA_DIR, \"clean\", \"pretransplant1.csv\"))\n",
    "posttransplant1[top_genes].to_csv(pjoin(DATA_DIR, \"clean\", \"posttransplant1.csv\"))\n",
    "\n",
    "pretransplant2[top_genes].to_csv(pjoin(DATA_DIR, \"clean\", \"pretransplant2.csv\"))\n",
    "posttransplant2[top_genes].to_csv(pjoin(DATA_DIR, \"clean\", \"posttransplant2.csv\"))\n",
    "\n",
    "healthy1[top_genes].to_csv(pjoin(DATA_DIR, \"clean\", \"healthy1.csv\"))\n",
    "healthy2[top_genes].to_csv(pjoin(DATA_DIR, \"clean\", \"healthy2.csv\"))\n",
    "# import ipdb; ipdb.set_trace()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the following code is from pcpca experiment single_cell_bmmc.py file: \n",
    "# https://github.com/andrewcharlesjones/pcpca/blob/main/experiments/realworld/scrnaseq/single_cell_bmmc.py\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from os.path import join as pjoin\n",
    "from scipy.io import mmread\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "\n",
    "DATA_DIR = \".../singlecell_bmmc\"\n",
    "N_COMPONENTS = 10\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    pretransplant2 = pd.read_csv(\n",
    "        pjoin(DATA_DIR, \"clean\", \"pretransplant2.csv\"), index_col=0\n",
    "    )\n",
    "    posttransplant2 = pd.read_csv(\n",
    "        pjoin(DATA_DIR, \"clean\", \"posttransplant2.csv\"), index_col=0\n",
    "    )\n",
    "\n",
    "    healthy1 = pd.read_csv(pjoin(DATA_DIR, \"clean\", \"healthy1.csv\"), index_col=0)\n",
    "    # healthy2 = pd.read_csv(pjoin(DATA_DIR, \"clean\", \"healthy2.csv\"), index_col=0)\n",
    "\n",
    "    # Background is made up of healthy cells\n",
    "    Y = healthy1.values  # pd.concat([healthy1, healthy2], axis=0).values\n",
    "\n",
    "    X = pd.concat([pretransplant2, posttransplant2], axis=0).values\n",
    "    X_labels = [\"Pretransplant\" for _ in range(pretransplant2.shape[0])]\n",
    "    X_labels.extend([\"Posttransplant\" for _ in range(posttransplant2.shape[0])])\n",
    "    X_labels = np.array(X_labels)\n",
    "    assert X_labels.shape[0] == X.shape[0]\n",
    "\n",
    "    # Standardize\n",
    "    Y -= Y.mean(0)  # background\n",
    "    Y /= Y.std(0)\n",
    "    # Y = Y.T\n",
    "    X -= X.mean(0)  # foreground\n",
    "    X /= X.std(0)\n",
    "    # X = X.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4501\n",
      "1985\n"
     ]
    }
   ],
   "source": [
    "# get the dimension of the foreground \n",
    "n, p = X.shape\n",
    "m, p = Y.shape\n",
    "print(n)\n",
    "print(m)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
